### DISCUSSION TITLE
org.telegram.telegrambots.bots.AbsSender#executeAsync throws exception

### DISCUSSION THREAD

**Discussion by n0mer** [rubenlagus/TelegramBots#1345] [General]
hello,

`org.telegram.telegrambots.bots.AbsSender#executeAsync` throws `org.telegram.telegrambots.exceptions.TelegramApiException` although there is callback function `org.telegram.telegrambots.updateshandlers.SentCallback#onException` present.

**Comment by bvn13** [comment]
more code please

**Comment by n0mer** [comment]
@bvn13 sorry, what code do you need? 

Here is signature of AbsSender#executeAsync: https://github.com/rubenlagus/TelegramBots/blob/master/telegrambots-meta/src/main/java/org/telegram/telegrambots/bots/AbsSender.java#L39 

```java
public <T extends Serializable, Method extends BotApiMethod<T>, Callback extends SentCallback<T>> void executeAsync(Method method, Callback callback) throws TelegramApiException
```
Second parameter - `Callback`; Callback has onException method.

So i'm wondering why not send exception thrown via that `onException`, w/out declaring that the whole `executeAsync` method throws TelegramApiExceptoin.

**Comment by rubenlagus** [comment]
It throws it because it has to validate that both parameters are not-null. If any of them is null, an exception is thrown right away.

**Comment by n0mer** [comment]
@rubenlagus but why this is not done in callback's onException method?

**Comment by rubenlagus** [comment]
@n0mer Because this exception is throw if you provide a null callback, in such a case, it is imposible to call onException method.

In addition, the callback is intended for error occurring during method execution, not for errors occurred before it is actually executed
